### ISSUE TITLE
Support test methods with Kotlin suspend modifier

### ISSUE THREAD

**Issue by IRus** [junit-team/junit5#1914]
## Goals

Support running suspend test in JUnit Jupiter:

```kotlin
class Kit {
    @Test 
    suspend fun foo() {
        delay(1000) // suspend call
        assertEquals(1, 1)
    } 
}
```

Currently, such test can be written this way:

```kotlin
class Kit {
    @Test 
    fun foo() = runBlocking {
        delay(1000) // suspend call
        assertEquals(1, 1)
        assertThrows { /* ... */ } 
        Unit // test should return void, but `assertThrows` returns `Throwable`, so `foo` returns `Throwable` too 
    } 
}
```

Also, will be nice to provide `CoroutineScope` through params, or as receiver in extension:

```kotlin
class Kit {
    suspend fun foo(scope: CoroutinesScope) {  /* ... */  } // (1)
    suspend fun CoroutinesScope.foo() {  /* ... */  } // (2)
}
```
1 and 2 actually the same on bytecode level. `suspend` is optional.

And finally, support for `runBlockingTest`:

```kotlin
class Kit {
    suspend fun foo(scope: TestCoroutinesScope) {  /* ... */  }
    suspend fun TestCoroutinesScope.foo() {  /* ... */  } 
}
```

## What can be done currently

`ParameterResolver` can be used to provide stubs for `Continuation`, `CoroutineScope` and `TestCoroutineScope`. These stub arguments can be replaced with real arguments in invocation.

## Problems

Current extensions points not enough to implement this feature as extensions, since:

1. **Discovery**. Jupiter discovers tests that returns `void`, but `suspend fun` returns `Object`;
2. **Invocation**. `InvocationInterceptor` in 5.5-M1(SNAPSHOT) don't providing mechanism to override actual invocation, only to decoration of existing invocation. Conversion of `method` to `kotlinFunction`, and then executing using [`callSuspend`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect.full/call-suspend.html) is necessary to execute `suspend fun`.

Also, my [slides](https://ruslan.ibragimov.by/pages/cv/speaker/kotlin-night-kiev-2019-junit.pdf) about this topic.

**Comment by sormuras** [comment]
Related to #1851

**Comment by marcphilipp** [comment]
@IRus This is not resolved by #2042, is it?

**Comment by IRus** [comment]
@marcphilipp Yes, not resolved, this is different issues

**Comment by xerus2000** [comment]
Please resolve this! My tests looked all fine until I discovered that they were not run at all anymore after adding suspending functions :/

**Comment by jmfayard** [comment]
This is a recurrent source of bugs for us.
A correct test that test a buggy piece of production code get ignored because of this.
Would be nice to see junit supports this.
In the meantime, I opeend a ticket to prevent IntelliJ users against that mistake

https://youtrack.jetbrains.com/issue/KT-52818/Provide-a-quick-fix-against-using-suspending-functions-in-Unit-Test

**Comment by ephemient** [comment]
I have discovered a way to make suspend work with JUnit Jupiter tests by bridging them through a `@TestTemplate`.
https://gist.github.com/ephemient/01d6e5766e6f8ea02839b4d7c3f94e55
However, this is not ideal as it doesn't play well with other extensions (e.g. #378).
There really needs to be either better extension points or built-in support for this.

**Comment by allertonm** [comment]
I also ran into this because I had the idea of trying to write a `TestMethodInvoker` that worked with the Quarkus test extensions that would handle setting up Vertx & Panache session context for suspend functions similar to the existing `@RunOnVertxContext` annotation that works with `Mutiny.Uni`. This would appear to be perfectly possible if it were not for JUnit ignoring non-void methods.

(I suspect this limitation also explains the cumbersome `UniAsserter` mechanism that `@RunOnVertxContext` uses, rather than having test methods that return Uni<Void>.)

**Comment by ianthetechie** [comment]
For anyone coming from a search engine, this is still not supported, but it looks possible using [`runTest`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-test.html), which replaces the [deprecated](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/MIGRATION.md) `runBlockingTest`.

**Comment by JavierSegoviaCordoba** [comment]
I am using it.

Adding `suspend` support is not straightforward as `runTest` accepts parameters and add a receiver to the scope, which would be missing by only adding `suspend` to the test.

**Comment by ianthetechie** [comment]
Yes runTest works fine. It’s just rather confusing that it’s not supported out of the box at this point.

**Comment by viluon** [comment]
`runTest` is great...

... until you need to integrate with [Lincheck](https://github.com/JetBrains/lincheck) and would like to run suspending set up code in `@BeforeAll`.

**Comment by marcphilipp** [comment]
There is now a draft PR for this issue:

- #4515 

I would very much welcome your feedback on the PR!

**Comment by marcphilipp** [comment]
Done in #4530

**Comment by marcphilipp** [comment]
You're referring to https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/run-test.html, right? I wasn't sure it was a good fit because it returns `TestResult` which is not consumable (unless you manually use `runTest`) and it skips delays which I thought would be rather surprising. It may be possible to convince me to change it, though, if you have good arguments. 🙂

**Comment by rgmz** [comment]
Thanks for adding this, @marcphilipp!

> ... and it skips delays which I thought would be rather surprising. 

IMO fast-forwarding delays is desirable behaviour. I don't want, or necessarily need, my ExponentialBackOff test cases to actually take 2/4/8/16/32/64/128... seconds.

**Comment by IRus** [comment]
Using `runTest` would imply passing `TestScope` to test functions somehow, otherwise user can't call something like https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/advance-time-by.html

Specifying `TestScope` is even more symbols than just writing `runTest`

Also this would imply that suspend tests would work only if kotlinx-coroutines-test provided by user or junit-platform-common should include it, which I guess not desirable

**Comment by opatry** [comment]
Also, having `TestScope` being provided to the test would allow using `advanceUntilIdle` and stuff like that which is very desirable also.

**Comment by ephemient** [comment]
I would have used `kotlin.coroutines.jvm.internal.runSuspend` instead of `kotlinx.coroutines.runBlocking` - it's in kotlin-stdlib-jvm so it's always present, and if a user wants a real `CoroutineScope` or `TestScope` they can use `runBlocking` or `runTest` themselves.

**Comment by JavierSegoviaCordoba** [comment]
@marcphilipp if I remember correctly, it is highly recommended to use "real" `delay` on tests instead of utility functions to add waits and/or move forward in time. Maybe @dkhalanskyjb can help with this, as I vaguely remember his comments.

`runBlocking` would make those tests too slow, so `runTest` is the only way to go. Additionally it is necessary to be able to change the dispatcher somehow.

It would be amazing if it were possible to set the dispatcher for the whole test class for example:

```kotlin
@TestDispatcher(FooDispatcher::class)
class BarTest {

    ...
}
```

Even I would find a bit weird to need to replace the dispatcher only on a single test instead of the whole class, `@TestDispatcher(FooDispatcher::class)` should work if it is annotating a test method too.

**Comment by marcphilipp** [comment]
@JavierSegoviaCordoba Could you please create a new issue for `runTest` and `@TestDispatcher`?

**Comment by JavierSegoviaCordoba** [comment]
@marcphilipp created here: https://github.com/junit-team/junit-framework/issues/4768

**Comment by IRus** [comment]
@marcphilipp  I think @ephemient having a very good point about using `runSuspend` (which is internal, but implementation is very straightforward and can be re-implemented easily).

`runBlocking` doing some clever things internally: https://madhead.me/posts/chm-not-concurrent-runblocking-not-blocking/

And `runSuspend` just blocking current thread until coroutine finished without any extra hidden behavior

**Comment by IRus** [comment]
Another nice property of `runSuspend` - it can be implemented without kotlinx.corountes, so JUnit wouldn't imply runtime dependency on kotlinx-coroutines-core. Which of course likely be present for people that want to use suspend tests, but not having plus one dependency is usually good thing

**Comment by marcphilipp** [comment]
Let's please continue the discussion in #4768
